File: app/src/main/kotlin/org/wordpress/aztec/demo/GlideImageLoader.java
package org.wordpress.aztec.demo;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.drawable.Drawable;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.Request;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SizeReadyCallback;
import com.bumptech.glide.request.target.Target;

import org.wordpress.aztec.Html;

import java.util.Locale;

public class GlideImageLoader implements Html.ImageGetter {

    public class LoggingListener<T, R> implements RequestListener<T, R> {
        @Override public boolean onException(Exception e, Object model, Target target, boolean isFirstResource) {
            android.util.Log.d("GLIDE", String.format(Locale.ROOT,
                    "onException(%s, %s, %s, %s)", e, model, target, isFirstResource), e);
            return false;
        }
        @Override public boolean onResourceReady(Object resource, Object model, Target target, boolean isFromMemoryCache, boolean isFirstResource) {
            android.util.Log.d("GLIDE", String.format(Locale.ROOT,
                    "onResourceReady(%s, %s, %s, %s, %s)", resource, model, target, isFromMemoryCache, isFirstResource));
            return false;
        }
    }

    private Context context;

    public GlideImageLoader(Context context) {
        this.context = context;
    }

    @Override
    public void loadImage(String source, final Callbacks callbacks, final int maxWidth) {
        Glide.with(context).load(source).fitCenter().listener(new LoggingListener<String, GlideDrawable>()).into(new Target<GlideDrawable>() {
            @Override
            public void onLoadStarted(Drawable placeholder) {
                Resources r = context.getResources();
            }

            @Override
            public void onLoadFailed(Exception e, Drawable errorDrawable) {
                callbacks.onImageLoadingFailed();
            }

            @Override
            public void onResourceReady(GlideDrawable resource, GlideAnimation<? super GlideDrawable> glideAnimation) {
                callbacks.onImageLoaded(resource);
            }

            @Override
            public void onLoadCleared(Drawable placeholder) {
            }

            @Override
            public void getSize(SizeReadyCallback cb) {
                cb.onSizeReady(maxWidth, Target.SIZE_ORIGINAL);
            }

            @Override
            public void setRequest(Request request) {

            }

            @Override
            public Request getRequest() {
                return null;
            }

            @Override
            public void onStart() {

            }

            @Override
            public void onStop() {

            }

            @Override
            public void onDestroy() {

            }
        });
    }
}

File: glide-loader/src/main/java/org/wordpress/aztec/glideloader/GlideImageLoader.kt
package org.wordpress.aztec.glideloader

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.DisplayMetrics
import com.bumptech.glide.Glide
import com.bumptech.glide.request.Request
import com.bumptech.glide.request.target.SizeReadyCallback
import com.bumptech.glide.request.target.Target
import com.bumptech.glide.request.transition.Transition
import org.wordpress.aztec.Html
import org.wordpress.aztec.glideloader.extensions.upscaleTo

class GlideImageLoader(private val context: Context) : Html.ImageGetter {

    override fun loadImage(source: String, callbacks: Html.ImageGetter.Callbacks, maxWidth: Int) {
        loadImage(source, callbacks, maxWidth, 0)
    }

    override fun loadImage(source: String, callbacks: Html.ImageGetter.Callbacks, maxWidth: Int, minWidth: Int) {
        Glide.with(context).asBitmap().load(source).into(object : Target<Bitmap> {
            override fun onLoadStarted(placeholder: Drawable?) {
                callbacks.onImageLoading(placeholder)
            }

            override fun onLoadFailed(errorDrawable: Drawable?) {
                callbacks.onImageFailed()
            }

            override fun onResourceReady(resource: Bitmap, transition: Transition<in Bitmap>?) {
                //Upscaling bitmap only for demonstration purposes.
                //This should probably be done somewhere more appropriate for Glide (?).
                if (resource.width < minWidth) {
                    return callbacks.onImageLoaded(BitmapDrawable(context.resources, resource.upscaleTo(minWidth)))
                }

                // By default, BitmapFactory.decodeFile sets the bitmap's density to the device default so, we need
                // to correctly set the input density to 160 ourselves.
                resource.density = DisplayMetrics.DENSITY_DEFAULT
                callbacks.onImageLoaded(BitmapDrawable(context.resources, resource))
            }

            override fun onLoadCleared(placeholder: Drawable?) {}

            override fun getSize(cb: SizeReadyCallback) {
                cb.onSizeReady(maxWidth, Target.SIZE_ORIGINAL)
            }

            override fun removeCallback(cb: SizeReadyCallback) {
            }

            override fun setRequest(request: Request?) {
            }

            override fun getRequest(): Request? {
                return null
            }

            override fun onStart() {
            }

            override fun onStop() {
            }

            override fun onDestroy() {
            }
        })
    }
}
